Enh/update v2.1

sanjayankur31 opened the pull request:
Regenerates the schemas for NeuroML v2.1.

Also updates the version of the library to 2.1.0: the MAJOR implies that it supports NeuroML 2.

sanjayankur31 committed:
enh: regenerate for NeuroML 2.1

sanjayankur31 committed:
chore: do not bundle xsd

sanjayankur31 committed:
enh: update Readme

sanjayankur31 committed:
enh: add tag/release badges

sanjayankur31 commented:
CC: @pgleeson

sanjayankur31 commented:
Note: needs to be migrated from travis-ci.org to something else now, either travis-ci.com or GitHub actions, or something else.

sanjayankur31 committed:
enh: update travis file

sanjayankur31 committed:
enh: use kitware ppa

We need a new version of cmake for the NAME_WLE options in
get_filename_component

sanjayankur31 commented:
Hrm, we need a new enough version of cmake, and I can't seem to get it on Ubuntu :roll_eyes: 

This is because we need to use `NAME_WLE` in `get_filename_component`: otherwise it thinks `.1.xsd` is the extension and doesn't compile correctly:

https://cmake.org/cmake/help/v3.14/command/get_filename_component.html

sanjayankur31 committed:
Use official CMake builds from Kitware

https://riptutorial.com/cmake/example/4723/configure-travis-ci-with-newest-cmake

sanjayankur31 committed:
Revert "Use official CMake builds from Kitware"

This reverts commit c28ba64d7a0cc149821ba6523c7d2435f7cc965a.

sanjayankur31 committed:
enh: use ubuntu 20.04 as build image on travis

sanjayankur31 committed:
enh: add github action for CI

sanjayankur31 committed:
Merge branch 'master' into enh/update-v2.1

sanjayankur31 committed:
Merge branch 'master' into enh/update-v2.1

sanjayankur31 commented:
@pgleeson @jonc125 : could you take a quick look at this when you have a minute? If everything looks OK, I'll merge and tag a release.

I'm using GH actions to also generate and push the API documentation to https://neuroml.github.io/NeuroML_API/. So we can just use GH actions instead of migrating from travis-ci.org to travis-ci.com

sanjayankur31 committed:
fix: make gh-pages force push

jonc125 commented on CMakeLists.txt:
```diff
@@ -74,11 +74,6 @@ if (XSD_FOUND)
                       COMMENT "Copying new XSD wrappers to source tree"
                       VERBATIM)
-else (XSD_FOUND)
-    # Use pre-generated sources & CodeSynthesis headers shipped with the project
```
Is there a reason to take this out? The readme still suggests having xsd installed is optional.

sanjayankur31 replied:
That's primarily so that a system copy of XSD is always used: which will receive updates from the repositories etc. Our bundled copy is a snapshot of the XSD code that will not be updated unless we remember to update it manually whenever there's a new upstream release. 

I'll update the readme to reflect this.

sanjayankur31 replied:
Ah, hold on. I messed this up. I didn't need to remove the whole section. Fixing now.

sanjayankur31 replied:
I've updated the CMakeLists.txt file and the Readme too. XSD is now a compulsory requirement, and a new option `REGENERATE_BINDINGS` can be called to generate the bindings. 

I've also made some tweaks to the doc generation bit.

jonc125 commented on README.md:
```diff
@@ -42,41 +44,50 @@ The NeuroML API is built using CMake, and the process works best with
 * `ctest` (or `make test`) to run tests,
 * `sudo make install` to install on your system, or
-* `cpack` to generate an installable package (see [cpack documentation]).
+* `cpack` to generate an installable package (see [cpack documentation](https://www.codesynthesis.com/products/xsd/)).
```
Wrong link here!

sanjayankur31 replied:
Fixed!

jonc125 commented on README.md:
```diff
@@ -42,41 +44,50 @@ The NeuroML API is built using CMake, and the process works best with
 
 The default CMake options should be appropriate for most cases. The only
-essential external dependency is [Xerces C++] which should be found
+essential external dependency is [Xerces C++](https://www.codesynthesis.com/products/xsd/) which should be found
```
And here!

sanjayankur31 replied:
Fixed!

jonc125 reviewed (commented):
Looks OK to me from a quick glance, apart from a couple of copy & paste errors. Just curious about removing the option to build without XSD installed.

sanjayankur31 committed:
fix: correct URLs

sanjayankur31 committed:
fix: correct instructions in Readme

sanjayankur31 committed:
chore: set minimum version of cmake required

sanjayankur31 committed:
fix: correct cmake project build logic

Since we do not bundle XSD any more, it is now a compulsory requirement
that the user must install themselves along with Xerces.

This also adds a new option `REGENERATE_BINDINGS` that can be used to
regenerate bindings.

sanjayankur31 committed:
enh: add metadata and logo to Doxygen docs

sanjayankur31 committed:
fix: update CI configuration to use new option for regenerating bindings

jonc125 reviewed (approved):
Looks good to me!

sanjayankur31 commented:
Thanks very much. I'll merge and tag a new release now.
